### ISSUE TITLE
[BUG] Archi 4.7 Saving  - Connection Relationship Errors

### ISSUE THREAD

**Issue by reed155** [archimatetool/archi-modelrepository-plugin#136]
[archi4_7Error.docx](https://github.com/archimatetool/archi/files/4471327/archi4_7Error.docx)

### Version of Archi
4.7 Early Access

### Archi Plug-ins
Archi Excel Plugin
Archi Scripting
Lightbox

### Operating System
Windows 10

### Expected Behaviour
Should save without error as in 4.6

### Actual Behaviour
Does not Save, presents an error.

### Steps to Reproduce the Behaviour
1. Edit model
2. Attempt to Save
3.

Please contact me privately of the error log if needed.

**Comment by Phillipus** [comment]
OK, thanks for the info. I've transferred this issue to the coArchi issue tracker under the assumption that this is the cause. This way we can investigate a solution based on @jbsarrodie comments:

> I think we should make sure when reloading the model that all connections are in sync with underlying relationship, and if not simply update (if new source and/or target are still on the same view) or delete them.

**Comment by Phillipus** [comment]
@reed155 But send me your model if you like so I can attempt a manual fix.

**Comment by michaelansley** [comment]
Hi,

I'm receiving the same error with very similar circumstances: Archi 4.6, coArchi, team of about 20 people using the model, branching, jArchi, and then towards the end of a merge process I get the missing folder.xml error message.  When I loaded the model in 4.7, I started receiving the error indicated by @reed155.  If I can provide any additional info or support, please let me know.

Michael

**Comment by michaelansley** [comment]
PS: All of the team are using coArchi 7.2, although I have just (today) installed coArchi 0.8.  It was with Archi 4.7/coArchi 0.8 that I received the error as indicated by @reed155.

**Comment by Phillipus** [comment]
Michael, I think you are referring to jArchi. coArchi is on version 0.6.2. jArchi won't have any bearing on this issue. What error messages are you seeing?

**Comment by jbsarrodie** [comment]
@michaelansley 

> I get the missing folder.xml error message

What you experience is not related to this issue but this one #130 (I suggest you read it carefully to understand the issue and its potential impacts with coArchi <0.6.2. The root cause is some views that have been both moved and notably updated by different people, leading Git to no more see the underlying view XML files as been the same. For the moment there is no real automatic solution. You have to find which views have been moved (the conflict dialog can help you in this) and then manually undo the move (which supposes that you still have the original folder in you model as it is identified by its internal ID, not its name). Once the move has been undo, you'll be able to merge your changes and publish them. You'll then be able to redo the move after.

>I started receiving the error indicated by @reed155

This one is related to this issue (and may or may not be related to coArchi btw). @Phillipus devised a small jArchi script to fix it:
````js
$('view').find('relationship').each(function(connection) {
    var relation = connection.concept;
    var relationSource = relation.source;
    var relationTarget = relation.target;
    var connectionSource = connection.source.concept;
    var connectionTarget = connection.target.concept;
    if(connectionSource.id != relationSource.id || connectionTarget.id != relationTarget.id) {
        connection.delete();
    }
});
````

This script will find nonvalid visual connection and remove them.

**Comment by Phillipus** [comment]
> This one is related to this issue (and may or may not be related to coArchi btw).

I wonder what else might cause this?

**Comment by jbsarrodie** [comment]
> I wonder what else might cause this? Could there be an edge case bug in reconnecting connections, for example?

TBH I do think this is caused by coArchi (as I can easily imagine a sequence of actions leading to this) but maybe there are some edge cases in jArchi (less likely though).

**Comment by michaelansley** [comment]
@Phillipus, of course, the 0.8 that I was referring to was jArchi, not coArchi, I mixed them up.  You are correct, I am using coArchi 0.6.2.

@jbsarrodie, thanks a lot for the info and the script reference.  I will check it and get back with info.

Have a great weekend!

**Comment by michaelansley** [comment]
Hi,

After checking the differences between the two branches, I decided to save the second branch as a file, switch back to the first branch and then use the new 4.7 functionality to import the file previously saved (of the second branch).  This imports the new views that are the critical changes from the second branch.

Once the file is imported to the first branch, I save.  Then I commit, after providing an appropriate message.  Then, when I try to Publish, I get an error message "There was an error <repository URL> 500 null".

I have switched to another branch and then switched back, same problem, although this time the error message is "Error writing request body to server"

I have tried the same procedure using a different but similar first branch, same problem.  Sometimes, I get the message "502 Bad gateway".

I have uploaded a text file with the console output: [Archi Error.txt](https://github.com/archimatetool/archi-modelrepository-plugin/files/4879943/Archi.Error.txt)

Is there any previous record of a successfully committed change causing a Publish error?  I've searched the issues but I haven't found anything.

**Comment by arnaudsp** [comment]
Hello, I have the same issue (Please see the #143). 
I want to apply the script to fix the issue.
Can you give me the access to JArchi Script to launch the script?

Best regards
